src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/TableMappingFactory.java

```diff
@@ -79,17 +87,108 @@ private void createTablesEagerly(CreateTableRequestFactory createTableRequestFac
+         */
+        void validateCompatiblePrimaryKeys(DynamoTableDescription virtualTable, DynamoTableDescription physicalTable) {
+            try {
+                partitioningStrategy.validateCompatiblePrimaryKeys(virtualTable.getPrimaryKey(),
+                        physicalTable.getPrimaryKey());
+            } catch (RuntimeException e) {
+                throw new IllegalArgumentException("incompatible table primary keys: "
+                        + e.getMessage() + ", virtualTable=" + virtualTable + ", physicalTable=" + physicalTable);
+            }
+        }
+
+        /**
+         * Validate that the physical table's primary key and all of its secondary index's primary keys have valid
+         * types.
+         */
+        void validatePhysicalTable(DynamoTableDescription physicalTableDescription) {
+            String tableMsgPrefix = "physical table " + physicalTableDescription.getTableName();
+            validatePhysicalPrimaryKey(physicalTableDescription.getPrimaryKey(), tableMsgPrefix);
+            physicalTableDescription.getGsis().forEach(dynamoSecondaryIndex ->
+                    validatePhysicalPrimaryKey(dynamoSecondaryIndex.getPrimaryKey(), tableMsgPrefix
+                            + "'s GSI " + dynamoSecondaryIndex.getIndexName()));
+            physicalTableDescription.getLsis().forEach(dynamoSecondaryIndex ->
+                    validatePhysicalPrimaryKey(dynamoSecondaryIndex.getPrimaryKey(), tableMsgPrefix
+                            + "'s LSI " + dynamoSecondaryIndex.getIndexName()));
+        }
+
+        private void validatePhysicalPrimaryKey(PrimaryKey primaryKey, String msgPrefix) {
+            checkArgument(partitioningStrategy.isPhysicalPrimaryKeyValid(primaryKey),
+                    msgPrefix + " has invalid primary key: " + primaryKey);
+        }
+    }
+
+    /**
      * Creates the table mapping, creates the table if it does not exist, sets the physical table description
      * back onto the table mapping so it includes things that can only be determined after the physical
      * table is created, like the streamArn.
      */
-    TableMapping getTableMapping(DynamoTableDescription virtualTableDescription) {
-        RandomPartitioningTableMapping tableMapping = new RandomPartitioningTableMapping(virtualTableDescription,
-            createTableRequestFactory,
-            secondaryIndexMapper,
-            mtContext);
-        tableMapping.setPhysicalTable(createTableIfNotExists(tableMapping.getPhysicalTable().getCreateTableRequest()));
+    TableMapping getTableMapping(DynamoTableDescription virtualTable) {
+        DynamoTableDescription physicalTable = lookupPhysicalTable(virtualTable);
+        physicalTable = createTableIfNotExists(physicalTable.getCreateTableRequest());
+        Map<DynamoSecondaryIndex, DynamoSecondaryIndex> secondaryIndexMap =
+                virtualTableCreationValidator.validateAndGetSecondaryIndexMap(virtualTable, physicalTable);
```

Comment:
When the index mapping is persisted, we'd get it from the repo record instead of generating it again here